integration/test/migration/migration_test.go

```diff
@@ -0,0 +1,367 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	longMaxInterval    = 2 * time.Minute
+	metricsAPIEndpoint = "/apis/metrics.k8s.io/v1beta1"
+	resourceNamespace  = metav1.NamespaceSystem
+	shortMaxInterval   = 5 * time.Second
+)
+
+// TestMigration ensures that previously deployed metrics-server chart will be deleted
+// before installing managed version of the metrics-server.
+func TestMigration(t *testing.T) {
+	ctx := context.Background()
+
+	// Install legacy resources.
+	err := helmClient.InstallFromTarball("/e2e/fixtures/resources-chart", resourceNamespace, helm.ReleaseName("resources"))
+	if err != nil {
+		t.Fatalf("could not install resources chart: %v", err)
+	}
+
+	// Check legacy resources are present.
+	err = checkLegacyResourcesPresent()
+	if err != nil {
+		t.Fatalf("legacy resources not present: %v", err)
+	}
+
+	// Check managed resources are not present.
+	err = checkManagedResourcesNotPresent("app=metrics-server,giantswarm.io/service-type=managed")
+	if err != nil {
+		t.Fatalf("managed resources present: %v", err)
+	}
+
+	// Install kubernetes-metrics-server-chart.
+	channel := fmt.Sprintf("%s-%s", env.CircleSHA(), testName)
+	err = r.Install(chartName, templates.MetricsServerValues, channel)
+	if err != nil {
+		t.Fatalf("could not install %q %v", chartName, err)
+	}
+
+	// Wait for deployed status
+	err = r.WaitForStatus(chartName, "DEPLOYED")
+	if err != nil {
+		t.Fatalf("timeout waiting for deployed resource: %v", err)
+	}
+
+	// Check legacy resources are not present.
+	err = checkLegacyResourcesNotPresent("app!=metrics-server,giantswarm.io/service-type!=managed")
+	if err != nil {
+		t.Fatalf("legacy resources present: %v", err)
+	}
+
+	// Check managed resources are present.
+	err = checkManagedResourcesPresent("app=metrics-server,giantswarm.io/service-type=managed")
+	if err != nil {
+		t.Fatalf("managed resources not present: %v", err)
+	}
+
+	// Check metrics availability
+	err = checkMetricsAvailability(ctx)
+	if err != nil {
+		t.Fatalf("could not get metrics: %v", err)
+	}
+
+	// Delete release
+	err = helmClient.DeleteRelease(chartName, helm.DeletePurge(true))
+	if err != nil {
+		t.Fatalf("failed to teardown resource: %v", err)
+	}
+}
+
+func checkMetricsAvailability(ctx context.Context) error {
+	var err error
+
+	restClient := h.K8sClient().CoreV1().RESTClient()
+
+	l.LogCtx(ctx, "level", "debug", "message", "waiting for the metrics become available")
+
+	o := func() error {
+
+		_, err := restClient.Get().RequestURI(metricsAPIEndpoint).Stream()
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+	b := backoff.NewConstant(longMaxInterval, shortMaxInterval)
+	n := backoff.NewNotifier(l, ctx)
+
+	err = backoff.RetryNotify(o, b, n)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	l.LogCtx(ctx, "level", "debug", "message", "successfully retrieved metrics from metrics server")
+
+	return nil
+}
+
+func checkLegacyResourcesPresent() error {
```

Comment:
Having separate functions for legacy and new resources makes this big. Is it possibly to set the `kind` labels we use in other tests?

https://github.com/giantswarm/kubernetes-kube-state-metrics/blob/afa4767283f4828cd13bc21da34fd3b0b8ad9a41/integration/test/migration/fixtures/resources-chart/templates/resources.yaml#L8

That allows you to use a label selector and just have functions for resources present or not present.

Reply:
Sure, I wanted to do it in that way. But the issue with it is - there are no labels on the rbac resources, only deployment and service. That's why it is impossible to distinguish those resources from others

Reply:
For the other charts we have a fixtures chart that creates the resources to delete. In that we set the kind label. 

I'm not sure if that works for metrics server. The benefit is it makes the test smaller but as it is now it works so not a blocker.

Reply:
There is no issue to set those labels in fixture. But in that case test itself doesn't repeat real state of things. I just put in fixtures real resources, which our customers are using

Reply:
Yes its true the labels don't exactly match. Although the kind label is ignored by the k8s-migrator job. This was what we ended up with for the previous charts. But if you prefer it this way that's fine.